File: lab4/menu.c
#include <stdio.h>
#include <stdlib.h>
#include "linktable.h"

void help();
void hello();
void quit();
void sum();
void compare();
void echo();
void ls();
void mkdir();

#define CMD_MAX_LEN 128
#define DESC_LEN    1024
#define CMD_NUM     128

typedef struct DataNode
{
    tLinkTableNode *pNext;
    char* cmd;
    char* desc;
    int   (*handler)();
    struct DataNode *next;
}tDataNode;
tDataNode *FindCmd(tLinkTable *head, char *cmd)
{
    tDataNode* pNode=( tDataNode *) GetLinkTableHead(head);
    while( pNode != NULL)
    {
        if( strcmp(pNode->cmd,cmd) == 0)
        {
            return pNode;
        }
        pNode=( tDataNode *) GetNextLinkTableNode(head,( tLinkTableNode *) pNode);
    }
    return NULL;
}
int ShowAllCmd( tLinkTable * head)
{
    tDataNode * pNode=( tDataNode *)GetLinkTableHead(head);
    while(pNode != NULL)
    {
        printf("%-10s - %s\n", pNode->cmd, pNode->desc);
        pNode =( tDataNode *) GetNextLinkTableNode( head, (tLinkTableNode *) pNode);
    }
    return 0;
}
int InitMenuData( tLinkTable ** ppLinkTable)
{
    * ppLinkTable = CreateLinkTable();
    tDataNode* pNode = ( tDataNode *) malloc( sizeof( tDataNode));
    pNode->cmd = "help";
    pNode->desc = "help command:";
    pNode->handler = help;
    AddLinkTableNode( * ppLinkTable,( tLinkTableNode *) pNode);
    pNode=( tDataNode *) malloc( sizeof( tDataNode));
    pNode->cmd = "hello";
    pNode->desc = "Say hello to users.";
    pNode->handler = hello;
    AddLinkTableNode(*ppLinkTable,(tLinkTableNode *)pNode);
    pNode = (tDataNode*)malloc(sizeof(tDataNode));
    pNode->cmd = "quit";
    pNode->desc = "Quit from Menu";
    pNode->handler = quit;
    AddLinkTableNode( * ppLinkTable,( tLinkTableNode *) pNode);
    pNode = ( tDataNode *) malloc( sizeof( tDataNode));
    pNode->cmd = "sum";
    pNode->desc = "An integer plus another";
    pNode->handler = sum;
    AddLinkTableNode(*ppLinkTable, (tLinkTableNode*) pNode);
    pNode = ( tDataNode *) malloc( sizeof( tDataNode));
    pNode->cmd = "compare";
    pNode->desc = "Get the bigger from two integers.";
    pNode->handler = compare;
    AddLinkTableNode(*ppLinkTable,(tLinkTableNode *)pNode);
    pNode = (tDataNode*)malloc(sizeof(tDataNode));
    pNode->cmd = "ls";
    pNode->desc = "List files in this dir.";
    pNode->handler = ls;
    AddLinkTableNode(*ppLinkTable,(tLinkTableNode *)pNode);
    pNode = (tDataNode*)malloc(sizeof(tDataNode));
    pNode->cmd = "echo";
    pNode->desc = "Output what you input.";
    pNode->handler = echo;
    AddLinkTableNode(*ppLinkTable,(tLinkTableNode *)pNode);
    pNode = (tDataNode*)malloc(sizeof(tDataNode));
    pNode->cmd = "mkdir";
    pNode->desc = "Make a new dir.";
    pNode->handler = mkdir;
    AddLinkTableNode(*ppLinkTable,(tLinkTableNode *)pNode);
    return 0;
}
tLinkTable * head= NULL;


int main()
{
    InitMenuData(&head);
    while(1)
    {
        char cmd[CMD_MAX_LEN];
        printf("Input a cmd: ");
        scanf("%s", cmd);
        tDataNode *p = FindCmd(head,cmd);
        if(p == NULL)
        {
            printf("WORRY COMMAND!\n");
            continue;
        }
        printf("%s - %s\n",p->cmd, p->desc);
        if(p->handler != NULL)
        {
            p->handler();
        }
    }
    return 0;
}

void help()
{
    ShowAllCmd(head);
}

void hello()
{ 
    printf("Hello ,Welcome to this menu~\n");
}

void quit()
{
    exit(0);
}

void sum()
{
    int a = 0;
    int b = 0;
 
    printf("Input two integers:\n");
    scanf("%d, %d", &a, &b);
    printf("a+b = %d\n", a+b);
}

void compare()
{
    int c = 0;
    int d = 0;
    int bigger = 0;

    printf("Input two integers:\n");
    scanf("%d, %d", &c, &d);
    bigger = c>d?c:d;
    printf("The bigger one is : %d\n", bigger);
}

void echo()
{
    char buf[128];
    printf("Start your performance:\n");
    fgetc(stdin);
    fputs(fgets(buf, 128, stdin), stdout);
}

void mkdir()
{
    printf("Make a new dir.\n");
}

void ls()
{
    system("ls");
}

File: lab5/go/menu.go
package main
import 
(
	"fmt"
	"os/exec"
	"os"
	"bufio"


)




var head *LinkTable = nil

func main()  {
	InitMenuData(&head)
	//ShowAllCmd(head)
	for true{
		var cmd string
		fmt.Printf("Input a cmd--> ")
		fmt.Scanln(&cmd)
		p := FindCmd(head, cmd)
		if p == nil{
			fmt.Println("WRONG COMMAND!!")
			continue
		}
		fmt.Println(p.cmd+" --- "+p.desc)
		
		handler(p)
		
	}
}

func hello() {
	fmt.Println("Hello, Wlecome to this menu~")
}
func help() {
	ShowAllCmd(head)
}
func quit() {
	os.Exit(1)
}
func sum() {
	a:=0
	b:=0

	fmt.Println("Input two numbers:")
	fmt.Scanf("%d %d", &a, &b)
	fmt.Println("Sum of two numbers: ",a+b)
}
func compare() {
	c:=0
	d:=0
	bigger:=0

    fmt.Println("Input two numbers:")
    fmt.Scanf("%d %d", &c, &d)
    if c>d {
    	bigger = c
    } else {
    	bigger = d
    }
    fmt.Println("The bigger one is: ", bigger)
}
func echo() {
	buf := bufio.NewReader(os.Stdin)

	line, err := buf.ReadString('\n')
	if err != nil{
		fmt.Println(err)
	}
	fmt.Println(line)
}
func mkdir() {
	fmt.Println("You want make a new dir? Sorry, i have no access to it.")
}
func ls() {
	cm := exec.Command("ls")
    out, err := cm.CombinedOutput()
    if err != nil {
        fmt.Println(err)
    }
    fmt.Println(string(out))
}
func handler(datanode *DataNode) {
	if datanode.cmd == "help" {
		help()
	}else if datanode.cmd == "hello" {
		hello()
	}else if datanode.cmd == "quit"{
		quit()
	}else if datanode.cmd == "sum"{
		sum()
	}else if datanode.cmd == "compare"{
		compare()
	}else if datanode.cmd == "echo"{
		echo()
	}else if datanode.cmd == "ls"{
		ls()
	}else if datanode.cmd == "mkdir"{
		mkdir()
	}
}
